### ISSUE TITLE
Variations stay stuck in OnBackorder status. Stock = 0

### ISSUE THREAD

**Issue by f3bruary** [woocommerce/woocommerce#19441]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [ ] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. A variable product has all variation's stock levels set to 0.
2. Product still appears to be On Backorder, even though it has actually sold out.
3. I viewed the product using WooCommerce Advanced Bulk Editor (CodeCanyon)
4. Here I see that the status of some variations are actually set to On Backorder instead of Out of Stock. The stock level is still 0.
5. When viewing the product in the product list view, pick quick edit, change status from On Backorder to Out of Stock, the changes won't persist and stay on On Backorder.
6. Bulk Edit plugin is also unable to change the status of these variations.

## Expected/actual behavior

When I follow those steps, I see the status changed won't persist.

I was expecting to see the status changing from On Backorder to Out of Stock.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

I don't know exactly after which WC update this started. I only just noticed it.

## WordPress Environment

<details>
```
### WordPress Environment ###

Home URL: 
Site URL: 
WC Version: 3.3.3
Log Directory Writable: ✔
WP Version: 4.9.4
WP Multisite: –
WP Memory Limit: 768 MB
WP Debug Mode: ✔
WP Cron: –
Language: nl_NL

### Server Environment ###

Server Info: Apache
PHP Version: 7.0.28
PHP Post Max Size: 128 MB
PHP Time Limit: 0
PHP Max Input Vars: 1000
cURL Version: 7.53.0
OpenSSL/1.0.2k

SUHOSIN Installed: –
MySQL Version: 5.6.36
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.3.3
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Totale databasegrootte: 122.26MB
Database datagrootte: 81.21MB
Database index grootte: 41.05MB
wp_usernamewoocommerce_sessions: Data: 0.08MB + Index: 0.02MB
wp_usernamewoocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_usernamewoocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_usernamewoocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_usernamewoocommerce_order_items: Data: 0.13MB + Index: 0.06MB
wp_usernamewoocommerce_order_itemmeta: Data: 1.52MB + Index: 1.78MB
wp_usernamewoocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_usernamewoocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
wp_usernamewoocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_usernamewoocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB
wp_usernamewoocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_usernamewoocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_usernamewoocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_usernamewoocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_usernameapsl_users_social_profile_details: Data: 0.02MB + Index: 0.03MB
wp_usernamearyo_activity_log: Data: 2.52MB + Index: 0.00MB
wp_usernameblc_filters: Data: 0.02MB + Index: 0.00MB
wp_usernameblc_instances: Data: 0.02MB + Index: 0.05MB
wp_usernameblc_links: Data: 0.02MB + Index: 0.06MB
wp_usernameblc_synch: Data: 0.16MB + Index: 0.09MB
wp_usernamecommentmeta: Data: 0.02MB + Index: 0.03MB
wp_usernamecomments: Data: 1.52MB + Index: 0.73MB
wp_usernamefailed_jobs: Data: 6.52MB + Index: 0.00MB
wp_usernameicwp_wpsf_ip_lists: Data: 0.02MB + Index: 0.00MB
wp_usernameicwp_wpsf_login_auth: Data: 0.02MB + Index: 0.00MB
wp_usernameicwp_wpsf_spambot_comments_filter: Data: 0.02MB + Index: 0.00MB
wp_usernameicwp_wpsf_statistics: Data: 0.02MB + Index: 0.00MB
wp_usernamekiyoh: Data: 0.02MB + Index: 0.00MB
wp_usernamelinks: Data: 0.02MB + Index: 0.02MB
wp_usernameloginizer_logs: Data: 0.00MB + Index: 0.01MB
wp_usernamemailchimp_carts: Data: 0.09MB + Index: 0.00MB
wp_usernamemollie_pending_payment: Data: 0.02MB + Index: 0.00MB
wp_usernamenf_objectmeta: Data: 0.02MB + Index: 0.00MB
wp_usernamenf_objects: Data: 0.02MB + Index: 0.00MB
wp_usernamenf_relationships: Data: 0.02MB + Index: 0.00MB
wp_usernameninja_forms_fav_fields: Data: 0.05MB + Index: 0.00MB
wp_usernameninja_forms_fields: Data: 0.02MB + Index: 0.00MB
wp_usernameoptions: Data: 9.20MB + Index: 2.20MB
wp_usernamepmxe_exports: Data: 0.02MB + Index: 0.00MB
wp_usernamepmxe_google_cats: Data: 0.39MB + Index: 0.00MB
wp_usernamepmxe_posts: Data: 0.02MB + Index: 0.00MB
wp_usernamepmxe_templates: Data: 0.02MB + Index: 0.00MB
wp_usernamepmxi_files: Data: 0.02MB + Index: 0.00MB
wp_usernamepmxi_history: Data: 0.02MB + Index: 0.00MB
wp_usernamepmxi_imports: Data: 0.09MB + Index: 0.00MB
wp_usernamepmxi_posts: Data: 0.06MB + Index: 0.00MB
wp_usernamepmxi_templates: Data: 0.02MB + Index: 0.00MB
wp_usernamepostmeta: Data: 43.59MB + Index: 29.92MB
wp_usernameposts: Data: 8.45MB + Index: 3.47MB
wp_usernamequeue: Data: 0.02MB + Index: 0.00MB
wp_usernamesocial_users: Data: 0.02MB + Index: 0.02MB
wp_usernametaxonomymeta: Data: 0.02MB + Index: 0.03MB
wp_usernametermmeta: Data: 0.05MB + Index: 0.03MB
wp_usernameterms: Data: 0.02MB + Index: 0.03MB
wp_usernameterm_relationships: Data: 2.38MB + Index: 1.39MB
wp_usernameterm_taxonomy: Data: 0.05MB + Index: 0.03MB
wp_usernameualp_user_activity: Data: 0.02MB + Index: 0.00MB
wp_usernameusermeta: Data: 0.44MB + Index: 0.25MB
wp_usernameusers: Data: 0.06MB + Index: 0.05MB
wp_usernamewcpdf_invoice_number: Data: 0.02MB + Index: 0.00MB
wp_usernamewc_download_log: Data: 0.02MB + Index: 0.03MB
wp_usernamewc_point_of_sale_sale_reports: Data: 0.00MB + Index: 0.00MB
wp_usernamewc_poin_of_sale_grids: Data: 0.02MB + Index: 0.00MB
wp_usernamewc_poin_of_sale_outlets: Data: 0.02MB + Index: 0.00MB
wp_usernamewc_poin_of_sale_receipts: Data: 0.02MB + Index: 0.00MB
wp_usernamewc_poin_of_sale_registers: Data: 0.02MB + Index: 0.00MB
wp_usernamewc_poin_of_sale_tiles: Data: 0.02MB + Index: 0.00MB
wp_usernamewc_webhooks: Data: 0.02MB + Index: 0.02MB
wp_usernamewebwinkelkeur_invite_error: Data: 0.02MB + Index: 0.03MB
wp_usernamewoof_query_cache: Data: 0.17MB + Index: 0.00MB
wp_usernamewoo_sr_cart_items: Data: 0.02MB + Index: 0.03MB
wp_usernamewoo_sr_orders: Data: 0.02MB + Index: 0.05MB
wp_usernamewoo_sr_orders_meta_all: Data: 0.02MB + Index: 0.00MB
wp_usernamewoo_sr_order_items: Data: 0.02MB + Index: 0.02MB
wp_usernamewpmelon_advbedit_temp: Data: 0.47MB + Index: 0.00MB
wp_usernamewpsp_agent_settings: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_attachments: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_canned_reply: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_catagories: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_custom_fields: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_custom_priority: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_custom_status: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_faq: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_faq_catagories: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_panel_custom_menu: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_ticket: Data: 0.02MB + Index: 0.00MB
wp_usernamewpsp_ticket_thread: Data: 0.02MB + Index: 0.00MB
wp_usernamewsluserscontacts: Data: 0.02MB + Index: 0.02MB
wp_usernamewslusersprofiles: Data: 0.02MB + Index: 0.03MB
wp_usernameyith_wcwl: Data: 0.02MB + Index: 0.02MB
wp_usernameyith_wcwl_lists: Data: 0.02MB + Index: 0.03MB
wp_usernameyoast_seo_links: Data: 0.38MB + Index: 0.14MB
wp_usernameyoast_seo_meta: Data: 1.50MB + Index: 0.00MB

### Post Type Counts ###

activity: 1
attachment: 3969
blocks: 31
bnfw_notification: 6
custom_css: 1
customize_changeset: 11
featured_item: 7
jetpack_migration: 2
mc4wp-form: 2
nav_menu_item: 69
nf_sub: 25
page: 43
pos_temp_register_or: 7
post: 5
product: 850
product_variation: 4659
ptp_variation: 7
pys_fb_event: 1
revision: 79
safecss: 1
shop_coupon: 480
shop_order: 2100
shop_order_refund: 114
sidebar: 1
wafs: 2
wpcf7_contact_form: 2
wpephpcompat_jobs: 39

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (51) ###

Activity Log: door Yakir Sitbon
Maor Chasen
Ariel Klikstein – 2.4.1

Breadcrumb NavXT: door John Havlik – 6.0.4
Broken Link Checker: door Janis Elsts
Vladimir Prelovac – 1.11.5

Buttonizer - Smart Floating Action Button: door Buttonizer – 1.4.4
Contact Form 7: door Takayuki Miyoshi – 5.0.1
Cookie Notice: door dFactory – 1.2.41
CSS Plus: door Paulo Calixto – 1.5.1
Custom Product Stickers voor Woocommerce: door starblank.com – 1.8.6 – Niet getest met de actieve versie van WooCommerce
Speciale Sidebars: door WPMU DEV – 3.1.2
Dashboard Widgets Suite: door Jeff Starr – 1.6
Google Tag Manager for Wordpress: door Thomas Geiger – 1.7.2
Easy Theme and Plugin Upgrades: door Chris Jean – 2.0.0
Error Log Monitor: door Janis Elsts – 1.5.7
Google Analytics Dashboard for WP (GADWP): door Alin Marcu – 5.3.1.1
Heartbeat Control: door Jeff Matson – 1.2.3
Kiyoh Customerreview: door kiyoh – 1.0.8
Loco Translate: door Tim Whitlock – 2.1.2
Loginizer: door Raj Kothari – 1.3.9
MailChimp for WordPress: door ibericode – 4.2
Mollie Payments for WooCommerce: door Mollie – 2.9.0
My Custom Functions: door Arthur Gareginyan – 4.13
Nextend Social Login: door Nextendweb – 3.0.4
PixelYourSite PRO: door PixelYourSite – 6.4.2
WooCommerce Product Filter: door Mihajlovic Nenad – 6.5.4 – Niet getest met de actieve versie van WooCommerce
Product Catalog Feed Pro by PixelYourSite: door PixelYourSite – 3.3.5.2
WooCommerce Pushover Integration: door Shop Plugins – 1.0.13
Regenerate Thumbnails: door Alex Mills (Viper007Bond) – 3.0.2
Search & Replace: door Inpsyde GmbH – 3.1.2
SG Optimizer: door SiteGround – 4.0.1
SiteOrigin CSS: door SiteOrigin – 1.1.5
Two Factor Authentication: door David Nutbourne + David Anderson
original plugin by Oskar Hane – 1.2.35 – Niet getest met de actieve versie van WooCommerce

User Rol Editor: door Vladimir Garagulya – 4.40.1
Woocommerce Advanced Bulk Edit: door George Iron – 4.3.4
WooCommerce Advanced Free Shipping: door Jeroen Sormani – 1.1.3 – Niet getest met de actieve versie van WooCommerce
WooCommerce DYMO Print Product Add-on: door WP Fortune – 3.1.8
WooCommerce DYMO Print: door WP Fortune – 3.1.8
Woocommerce Mailchimp Discount: door Magnigenie – 2.0 – Niet getest met de actieve versie van WooCommerce
WooCommerce MyParcel: door Richard Perdaan – 2.5.2 – Niet getest met de actieve versie van WooCommerce
WooCommerce PDF Invoices & Packing Slips: door Ewout Fernhout – 2.1.5
WooCommerce Point of Sale: door Actuality Extensions – 4.2.7.4
WooCommerce Product SKU Generator: door SkyVerge – 2.3.3 – Niet getest met de actieve versie van WooCommerce
WooCommerce: door Automattic – 3.3.3
WooSidebars: door WooThemes – 1.4.3
WordPress Importeerder: door wordpressdotorg – 0.6.4
Yoast SEO: door Team Yoast – 7.0.3
WP Instagram Widget: door Scott Evans – 2.0.3
WP Migrate DB: door Delicious Brains – 1.0.2
WP-Optimize: door David Anderson
Ruhani Rabin
Team Updraft – 2.2.2

WP Fortune: door WP Fortune – 1.0.8
Woocommerce Google Cloud Print: door Xperts Club – 1.5
YITH WooCommerce Wishlist: door YITHEMES – 2.2.1

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
heren (heren)
simple (simple)
variable (variable)
veterschoenen (veterschoenen)
zwart (zwart)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Winkelbasis: #2910 - /winkel/
Winkelmand: #2911 - /winkelmand/
Afrekenen: #2912 - /afrekenen/
Mijn account: #2913 - /mijn-account/
Algemene voorwaarden: #663 - /algemene-voorwaarden/

### Theme ###

Name: Flatsome
Version: 3.5.2
Author URL: http://www.uxthemes.com/
Child Theme: ❌ – Als je wijzigingen aanbrengt in WooCommerce of in een hoofdthema dat je niet zelf gebouwd hebt
raden we het gebruik van een child thema aan. Zie: Hoe maak je een child thema

WooCommerce Support: ✔

### Templates ###

Overrides: flatsome/woocommerce/archive-product.php
flatsome/woocommerce/cart/cart-empty.php
flatsome/woocommerce/cart/cart.php
flatsome/woocommerce/checkout/form-checkout.php
flatsome/woocommerce/checkout/form-coupon.php
flatsome/woocommerce/checkout/thankyou.php
flatsome/woocommerce/content-product.php
flatsome/woocommerce/content-product_cat.php
flatsome/woocommerce/content-single-product.php
flatsome/woocommerce/global/breadcrumb.php
flatsome/woocommerce/global/quantity-input.php
flatsome/woocommerce/global/wrapper-end.php
flatsome/woocommerce/global/wrapper-start.php
flatsome/woocommerce/loop/loop-end.php
flatsome/woocommerce/loop/loop-start.php
flatsome/woocommerce/loop/pagination.php
flatsome/woocommerce/loop/result-count.php
flatsome/woocommerce/loop/sale-flash.php
flatsome/woocommerce/myaccount/form-login.php
flatsome/woocommerce/myaccount/navigation.php
flatsome/woocommerce/notices/error.php
flatsome/woocommerce/notices/notice.php
flatsome/woocommerce/notices/success.php
flatsome/woocommerce/product-searchform.php
flatsome/woocommerce/single-product/photoswipe.php
flatsome/woocommerce/single-product/price.php
flatsome/woocommerce/single-product/product-image.php
flatsome/woocommerce/single-product/product-thumbnails.php
flatsome/woocommerce/single-product/related.php
flatsome/woocommerce/single-product/review.php
flatsome/woocommerce/single-product/sale-flash.php
flatsome/woocommerce/single-product/share.php
flatsome/woocommerce/single-product/short-description.php
flatsome/woocommerce/single-product/tabs/tabs.php
flatsome/woocommerce/single-product/title.php
flatsome/woocommerce/single-product/up-sells.php
flatsome/woocommerce/single-product-reviews.php
flatsome/woocommerce/single-product.php
```
</details>

**Comment by thomas-zg** [comment]
How do you setup the [Allow backorders?] field in your products? If they are set to Allow, you'll see the status On backorder if stock level is 0.

**Comment by mikejolley** [comment]
This status is set automatically once stock is <= 0. If you want it out of stock, disable backorders for the variation/product. 

^ `Allow backorders?` is the option you'd need to disable.

**Comment by f3bruary** [comment]
The stock level for each variation is 0. Backorders are disabled for each variation.

**Comment by mikejolley** [comment]
Were they on backorder previously then?

**Comment by f3bruary** [comment]
Not likely. I have Allow Backorders disabled by default. I don't work with backorders. It's possible that this was enabled for maybe 1 or two products, but I have this issue for multiple products.

Even if they were, this should still allow me to set them to Out of Stock.

**Comment by mikejolley** [comment]
Well it's set automatically under certain conditions.  @claudiulodro Do we have a test case for 0 stock, not on backorder?

**Comment by claudiulodro** [comment]
@mikejolley [we have tests for all the possible combinations, including that one](https://github.com/woocommerce/woocommerce/blob/master/tests/unit-tests/product/data.php#L128-L134).

@f3bruary one thing that may be worth checking is if the parent variable product has stock management enabled and is backordered, [as the backorder status will propagate to the children](https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-variable-data-store-cpt.php#L490)

**Comment by f3bruary** [comment]
@claudiulodro parent has stock management disabled. Can variations still inherit this status when disabled?

To be clear, this never happened pre WC3

**Comment by claudiulodro** [comment]
@f3bruary The status syncing only happens if the parent variable product has stock management enabled, so it's not that.

**Comment by f3bruary** [comment]
Any other tips ? One of my variable products went out of stock this morning. Status is again stuck on 'On Backorder'. Stock management is only enabled for the variations. Stock level is 0. Stock status in Edit Product page is 'Out of Stock'. Stock status on product list page is 'On Backorder' and cannot be changed.

**Comment by f3bruary** [comment]
I fixed this by using a bulk editor and setting all variations' "Allow Backorders" to "No". The database wasn't reflecting this although the wp backend was. After bulk updating the variations they appear as 'Out of Stock' again.

**Comment by dustinpitcher** [comment]
I realize this is a closed topic but I don't see any resolution and this is causing repeated headaches for myself and customers. I've even attempted to use the 'WooCommerce_update_product" hook to catch erroneous "onbackorder" statuses and fix them right then but even that doesn't work. I have to do as @f3bruary said, change "allow backorders" to "Allow" then back to "Do Not Allow" and then the variants will begin showing "outofstock" again. If we only had a few products maybe it wouldn't be an issue. But with over 3000 products it is nearly impossible for us to catch these before they end up frustrating a customer. Because when this situation happens, what the customer sees is the product showing as available to purchase. They see an "Add to Cart" button, but when they hit the button, it doesn't allow them to add it to the cart. Needless to say, this is something that will bounce a potential customer. Love WooCommerce, love the team of contributors, hoping to get some love on this issue. Thanks all! -

**Comment by salsero72** [comment]
Exact same issue for me.
I sync (insert/update) products from another ecommerce using my own plugin 
Parent product -> manage stock = FALSE
Backorders set to "not allow" in each variation.

If variation stock_qty is >0 then everything is fine.
If variation stock_qty is = 0 then stock_status is set to "onbackorder" and there is no way to change using php.
By modifing the product and reset to it's original state, **onbackorder** is switched to **outofstock**

I even tried to save and resave with no luck.

`
    $variation->save();    $variation->set_stock_status('outofstock');
      $variation->save(); 
`

This happens with products that have some variations with mixed stock quantities (zero and more than zero) as well.

**Comment by salsero72** [comment]
Yeah! I SOLVED

For me the problem here was that I created variation without explicitly setting backorders to "not allowed".
So maybe woocommerce handle the product as if allows backorders.

So I just added 
`$variation->set_backorders('no');`

after creating the variation instance.

    // Get an instance of the WC_Product_Variation object
    $variation = new WC_Product_Variation( $variation_id );
    $variation->set_backorders('no');


then set the stock quantity to zero and stock status to outofstock
$variation->set_stock_quantity(0);
 $variation->set_stock_status('outofstock');

(sorry I can't wrap the code in code tag,. it seems it gets messed up in preview.)
